### ISSUE TITLE
Add new add-on command to wait for build to complete

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#641]
We should add a new addon command which allows for a build to complete and a service be accessible. It is not enough to just tail the build, instead one need to determine when a pod switches into the running state or in case of readiness checks defined as part of the application until the readiness checks pass.

**Comment by hferentschik** [comment]
See also https://github.com/openshift/jenkins-client-plugin

**Comment by jorgemoralespou** [comment]
@hferentschik 
I would have a waitForService and a waitForDeployment

waitForService can wait until an endpoint is > 0 (there is an actual endpoint that can be used)
waitForDeployment waits for a pod of that particular deployment to be Running

**Comment by hferentschik** [comment]
> waitForService and a waitForDeployment

sounds good. I like the names of the commands. Maybe we can unify a bit and have 'wait --forService' and 'wait --forDeployment'. If we then rename 'sleep' to 'wait', we have some sort of wait related group of commands. WDYT?

What would be the parameter for these new commands? The OpenShift service/deployment config names?

> waitForService can wait until an endpoint is > 0 

What do you mean exactly? How do you determine whether the endpoint "is > 0". Are you thinking of extracting  the readiness probe URL and using that? Or are you talking about a OpenShift API we should use. I kind of like the idea to extract and monitor the readiness probe where possible

> waitForDeployment waits for a pod of that particular deployment to be Running

+1

I think this is the most important new feature from the set of changes we discussed. This will open up new possibilities and enable doing proper deployments of applications. A lot of the other issues we discussed are more of cosmetic nature. I definitely want this for the next version.

**Comment by LalatenduMohanty** [comment]
> I like the names of the commands. Maybe we can unify a bit and have 'wait --forService' and 'wait --forDeployment'. If we then rename 'sleep' to 'wait', we have some sort of wait related group of commands. WDYT?

+1 sounds good to me.

**Comment by jorgemoralespou** [comment]
> waitForService can wait until an endpoint is > 0
When there is an endpoint you get a value for this expression. This can be queried to be different from null.

````
oc get endpoints/test -o jsonpath='{.subsets[*].addresses[*].ip}'
````

> waitForDeployment waits for a pod of that particular deployment to be Running

For this, you can monitor the pod status for a specific deploymentconfig 

Not working, but should be something similar.
````
oc get pods -o jsontemplate=".items[?(@.metadata.labels.deploymentconfig=='test')]].status.phase=='Running'"
````

Asking for help.

@stevekuznetsov Could you help us figure out how to determine if a deployment has finished (has all scheduled pods running, or at least 1)

**Comment by stevekuznetsov** [comment]
We have an endpoint for waiting for deployment status:

```
$ oc rollout status deploymentconfig NAME --watch
```

For builds, you can just tail the log:

```
$ oc logs buildconfig NAME --follow
```

**Comment by jorgemoralespou** [comment]
@stevekuznetsov this is great. By the way, that endpoint does not support "-o flag".

I guess this would be your recommendation to watch for a deployment being finished, rather than looking at the number of endpoints in the service.

**Comment by stevekuznetsov** [comment]
Sorry, where were you trying to use `-o` that didn't accept it? I'm not clear on what you meant.

The number of endpoints is a very indirect way to check for this -- what happens if someone edits the number of replicas while you're waiting? You might never return or you could return when it's not really ready yet. `oc rollout status` handles these sorts of cases.

**Comment by jorgemoralespou** [comment]
@stevekuznetsov Here:

````
oc rollout status deploymentconfig NAME -o json
````

**Comment by hferentschik** [comment]
good stuff, thanks. That helps a lot.

**Comment by stevekuznetsov** [comment]
A good thing to remember too with `oc logs --follow` is that it can time out waiting for the build to start, so you realistically need to check the phase, which has the following options:

```go
// BuildPhase represents the status of a build at a point in time.
type BuildPhase string

// Valid values for BuildPhase.
const (
	// BuildPhaseNew is automatically assigned to a newly created build.
	BuildPhaseNew BuildPhase = "New"

	// BuildPhasePending indicates that a pod name has been assigned and a build is
	// about to start running.
	BuildPhasePending BuildPhase = "Pending"

	// BuildPhaseRunning indicates that a pod has been created and a build is running.
	BuildPhaseRunning BuildPhase = "Running"

	// BuildPhaseComplete indicates that a build has been successful.
	BuildPhaseComplete BuildPhase = "Complete"

	// BuildPhaseFailed indicates that a build has executed and failed.
	BuildPhaseFailed BuildPhase = "Failed"

	// BuildPhaseError indicates that an error prevented the build from executing.
	BuildPhaseError BuildPhase = "Error"

	// BuildPhaseCancelled indicates that a running/pending build was stopped from executing.
	BuildPhaseCancelled BuildPhase = "Cancelled"
)
```

```sh
function wait-for-build() {
    local build_name=$1
    while true; do
        status=$( oc get build "${build_name}" -o jsonpath='{ .status.phase }' )
        if [[ "${status}" == "New" ||  "${status}" == "Pending" ||  "${status}" == "Running" ]]; then
            sleep 10
        else if [[  "${status}" == "Complete" ]]; then
            return 0
        else if [[  "${status}" == "Failed" ||  "${status}" == "Error" ||  "${status}" == "Cancelled" ]]; then
            return 1
        fi
    done
}

**Comment by hferentschik** [comment]
On second thought, I think we should indeed add the top level _waitForService_ and _waitForDeployment_ commands. This fits better into the current mini DSL approach and it will also be easier to document and in fact implement.

**Comment by stevekuznetsov** [comment]
What do you mean by waiting for a service?

**Comment by LalatenduMohanty** [comment]
`$ oc rollout status deploymentconfig NAME --watch` sounds perfect for waiting for build to complete.  I think this is enough for the addon command waitForDeployment. 
e.g. 

```
 oc rollout status deploymentconfig cakephp-ex --watch
Deployment config "cakephp-ex" waiting on image update
216734
Waiting for latest deployment config spec to be observed by the controller loop...
Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for rollout to finish: 0 of 1 updated replicas are available...
Waiting for latest deployment config spec to be observed by the controller loop...
Waiting for latest deployment config spec to be observed by the controller loop...
replication controller "cakephp-ex-1" successfully rolled out
```

**Comment by hferentschik** [comment]
> What do you mean by waiting for a service?

My understanding was that this means to wait until an application service, aka route is available. In my world this translates to monitoring a route and waiting until this route returns a 200. @jorgemoralespou, is this what you had in mind?

**Comment by LalatenduMohanty** [comment]
> What do you mean by waiting for a service?

@stevekuznetsov  @hferentschik I think this useful when we are trying to run multiple micro services and one service is waiting for another service. So the addon will wait till the dependent service is running and then deploy the other service.

**Comment by jorgemoralespou** [comment]
@hferentschik watching for a route is not useful. An application is ready once a deployment is ready. The only caveat could be if a service has not been created for the deployment, which would make the application unaccesible. 
We can watch for a route, but if you create a route and a deployment, watching should be for the deployment. The route get created anyway.

As @stevekuznetsov mentions, there should be no value in monitoring services by itself (only if you scale up/down can have additional value), and for the DSL for minishift should make no difference, so I would vote to have just 2 single action:

waitForDeployment

**Comment by stevekuznetsov** [comment]
Waiting for an application to be ready is a pretty thorny thing -- especially if you have multiple services working together to make it happen. What if your webserver is up and going but the database is still booting? You can hit the API of your app and it will respond... with 500. If a user wants to wait for their app to be ready they need to write their own code to determine what it is, exactly, that readiness means and how they are going to determine it. Maybe there's some use for a DSL around polling but Kubernetes and OpenShift don't provide any higher-level "all good" indicator.

**Comment by hferentschik** [comment]
> watching for a route is not useful. An application is ready once a deployment is ready. 

Maybe waiting for a route was describing it the wrong way. I was referring to the case we discussed in our meeting, where a pod might switch to running, but the app is still not ready, since it also defines a readiness probe. Maybe this is included in `oc rollout status deploymentconfig cakephp-ex --watch`? How does this rollout command behave in the case of build w/ resp w/o readiness probe?

> so I would vote to have just 2 single action: waitForDeployment

right, assuming it covers the above mentioned case, right?

**Comment by jorgemoralespou** [comment]
Openshift will use probes if they are defined. This is the best it can do to know if the ago is running. Oc rollout should use the probes if they are defined, and since this endpoint exists (which I didn't think of, when we talked) I silk consider just using this.

**Comment by LalatenduMohanty** [comment]
@hferentschik @jorgemoralespou Even if I am almost ready with the code, the scope is narrowed down to just one command i.e. `$ oc rollout status deploymentconfig nodejs-ex` . I tested this without `--watch` and still waits for the rollout to complete.  Please check the examples below.

I feel this is something you can use easily with the existing `oc` addon command. If it would have been complicated than this than it make more sense to create a separate command it.

```
$ oc rollout status deploymentconfig nodejs-ex
Deployment config "nodejs-ex" waiting on image update
Waiting for latest deployment config spec to be observed by the controller loop...
Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
Waiting for rollout to finish: 0 of 1 updated replicas are available...
Waiting for latest deployment config spec to be observed by the controller loop...
Waiting for latest deployment config spec to be observed by the controller loop...
replication controller "nodejs-ex-1" successfully rolled out

$ oc rollout status deploymentconfig nodejs-ex
replication controller "nodejs-ex-1" successfully rolled out

$ oc rollout status deploymentconfig nodejs-ex --watch
replication controller "nodejs-ex-1" successfully rolled out
```

**Comment by LalatenduMohanty** [comment]
I tried with a simple addon as below and it works fine
```
# Name: Deploy-nodejs
  2 # Description: Deploy Nodejs example and test waitForDeployment command
  3 # URL: https://github.com/openshift/nodejs-ex
  
sleep 10
oc new-app https://github.com/openshift/nodejs-ex -l name=myapp
oc rollout status deploymentconfig nodejs-ex
```

**Comment by hferentschik** [comment]
@LalatenduMohanty, you might be onto something. So here is the thing. For me the reason to have something like _waitForDeployment_ was that I did not have a way to create and app and wait for the app to be build/deployed, without akwards sleeps or wait loop.

What I used to do in my (shell) scripts was:

        $ oc new-app https://github.com/openshift/nodejs-ex -l name=myapp
        $ oc logs buildconfig nodejs-ex --follow

This would not work properly, since there is a timing issue. The tailing of the buildconfig won't work until the build actually started, which if the builder image is not there, can take some time. So effectively I would put the 'oc logs' command in a loop and try several times to start tailing. Either eventually it would work or after a certain time I would just break the loop. 

This is what I wanted to abstract the user from and which I thought _waitForDeployment_  was about.

So now, _oc rollout_ comes into play. Apparently I can do:

     $ oc new-app https://github.com/openshift/nodejs-ex -l name=myapp
     $ oc rollout status deploymentconfig nodejs-ex

Provided this has not timing issues like the 'oc logs' approach, the question becomes:" Do we really need a new command for the Minishift add-on DSL. As @LalatenduMohanty is saying a user can just add this command directly to the add-on. This is better than having to add a Minishift specific command, since we are effectively also educating the user. 

So, do we need a _waitForDeployment_ or not? @jorgemoralespou, @stevekuznetsov WDYT?

**Comment by hferentschik** [comment]
Independent of we add a new command or use the existing _oc_ add-on command, I think we need to make sure that the command executes in the separate go routine. Waiting for an application to be deployed can take a long time. I think we need to be able to show some sort of progress, otherwise the user might just think that the execution hangs. This is in particular true, if we disable the logging/printing of the commands executed.

**Comment by jorgemoralespou** [comment]
@hferentschik plus you need to properly handle errors in this scenario. I would not find helpful to get awkward output that I would not understand as an addon user. I would rather have the addon execution go, and have messages for failure at certain points, so if a deployment fails, I just get a "xyz deployment failed. Try again" or maybe a "xyz deployment failed, don't try again as it will always fail".

Exposing the underlying messages of __oc__ to the addons users are probably not helpful, and this should only show if loglevel>0 so he can send a report in an issue. Else, human readable errors are a must for me. So as long as you can achieve that with "oc rollout status dc/nodejs-ex" I'm fine, but I think this was my point for introducing a routine.

Also, showing progress might be helpful, but not sure how to show progress consistently for the overall addon execution, and not only for this step.

**Comment by hferentschik** [comment]
>  plus you need to properly handle errors in this scenario. I would not find helpful to get awkward output that I would not understand as an addon user. I would rather have the addon execution go, and have messages for failure at certain points, so if a deployment fails, I just get a "xyz deployment failed. Try again" or maybe a "xyz deployment failed, don't try again as it will always fail".

So right, improved error handling could be an argument pro dedicated command. Also maybe the ability to specify a max timeout. We discussed this as well which by the way also requires us to run the actual _oc_ command in a go routine. 

> Exposing the underlying messages of oc to the addons users are probably not helpful, and this should only show if loglevel>0 so he can send a report in an issue. Else, human readable errors are a must for me

Right we discussed this. I think you suggested that one can define a default error message or even an error message per command which prints on failure (and potentially a success message). Sounds reasonable to me, but I would consider this another issue.

> Also, showing progress might be helpful, but not sure how to show progress consistently for the overall addon execution, and not only for this step.

I have some ideas. I like how for example Minitest or Cucumber to it in the "quiet" output. Basically they add a '.' for each test executed. For us this would mean, printing just a single dot for each command executed, for example:

    $ minishift start 
    ...
    Applying add-on foo .......

Now, when a command starts to take longer than a few second we start putting up a spinner (rotating -\|/) in the way go-spin does for example. Once the command succeeds we print the '.'. In the end yu end up with something like

    Applying addon-on foo ........... success

Where any of the '.' might have been spinning for some time.

**Comment by stevekuznetsov** [comment]
> I would not find helpful to get awkward output that I would not understand as an addon user.

Unless you think there is a significant difference between the set of users that run `oc` and the set of users for your platform, issues with the output should be fixed with pull requests to Origin.

**Comment by praveenkumar** [comment]
@hferentschik any reason we removed a milestone for it?

**Comment by hferentschik** [comment]
Because it is not really needed due to _oc rollout_. I think we can actually close this one.
